guides/testing: Add testing section (#333)
unikraft/docs: content/guides/testing.mdx

---
title: Testing Unikraft
description: |
  We are going to explore the idea of validation by testing.
  The main focus will be testing but we also tacke other validation methods such as fuzzing and symbolic execution.

>   Expected time: 15 minutes

Reviewer: ```suggestion
```
Remove this, it's clearly not 15 mins.

* **Branch coverage** is the percentage of branches executed during the testing (e.g. if or while).
* **Path coverage** is the percentage of paths executed during the testing.

We'll now go briefly over two other validation techniques, fuzzing and symbolic execution.

### Fuzzing

**Fuzzing** or fuzz testing is an automated software testing technique that involves providing invalid, unexpected, or random data as inputs to a computer program.
The program is then monitored for exceptions such as crashes, failing built-in code assertions, or potential memory leaks.

The most popular OS fuzzers are [`kAFL`](https://github.com/IntelLabs/kAFL) and [`syzkaller`](https://github.com/google/syzkaller), but research in this area is very active.

### Symbolic Execution

As per Wikipedia, **symbolic execution** is a means of analyzing a program to determine what inputs cause each part of a program to execute.
An interpreter follows the program, assuming symbolic values for inputs rather than obtaining actual inputs as normal execution of the program would.
An example of a program being symbolically executed can be seen in the figure below:

> ![Symbolic execution](./images/symbex.png)

Reviewer: Use `<Image...` syntax. See an example [here](https://github.com/unikraft/docs/blob/main/content/docs/contributing/review-process.mdx#L59), the images should be placed under `static/assets/`.

KUnit is not the only tool used for testing Linux, there are tens of tools used to test Linux at any time:

* Test suites
  * [Linux Test Project](https://github.com/linux-test-project/ltp) is a collection of tools
  * Static code analyzers ([`Coverity`](https://scan.coverity.com/), [`Coccinelle`](https://coccinelle.gitlabpages.inria.fr/website/), [`smatch`](https://lwn.net/Articles/691882/), [`sparse`](https://sparse.docs.kernel.org/en/latest/))
  * Module tests (KUnit)
  * Fuzzing tools ([`Trinity`](https://github.com/kernelslacker/trinity), [`Syzkaller`](https://github.com/google/syzkaller))
  * Subsystem tests
* Automatic testing
  * `kisskb`
  * [`0Day`](https://github.com/0day-ci/linux)
  * [`kernelci`](https://foundation.kernelci.org/)
  * `Kerneltests`

In the figure below, we can see that as more and better tools were developed we saw an increase in reported vulnerabilities.
There was a peak in 2017, after which a steady decrease which may be caused by the amount of tools used to verify patches before being upstreamed.

> ![linux vulnerabilities](./images/linux_vulnerabilities.png)

Reviewer: Same as above, use `<Image...`, place images under `static/assets/`.

        uk_initcall_class_prio(UK_TESTSUITE_FN(suite), class, prio)
```

We can see that `uk_initcall_class_prio` registers the newly defined function as a constructor to be called at a specific time during the boot process.
`uk_testsuite_add` simply adds the test suite to a linked listed of available test suites.
`uk_testsuite_run` simply iterates runs all the test cases in the test suite.

```C++
int uk_testsuite_run(struct uk_testsuite *suite)
{
        ...
        /* Iterate through all the registered test cases */
        uk_testsuite_for_each_case(suite, testcase) {

        /* Run the test case function testcase->func(testcase);

                /* If one case fails, the whole suite fails. */
                if (testcase->failed_asserts > 0)
                        suite->failed_cases++;
        }
        ...
}
```

## Work Items

In this work session we will go over writing and running tests for Unikraft.
We will use `uktest` and [`Google Test`](https://github.com/google/googletest).
Make sure you are on the `usoc21` branch on the core Unikraft repo and `staging` on all others.
`uktest` should be enabled from the Kconfig.

### Support Files

> Session support files are available [on Google Drive](https://drive.google.com/drive/folders/1PEDtvR5W-eMGQgbJusMcDPc6hS6ghYAl?usp=sharing).

Reviewer: We should move the items under the [exercises repo](https://github.com/unikraft-upb/guides-exercises).

    UK_TEST_EXPECT_SNUM_EQ(ret, 0);

    ret = mkdir("/foo", S_IRWXU);
    UK_TEST_EXPECT_SNUM_EQ(ret, 0);

    fd = open("/foo/file.txt", O_WRONLY | O_CREAT, S_IRWXU);
    UK_TEST_EXPECT_SNUM_GT(fd, 2);

    UK_TEST_EXPECT_SNUM_EQ(
        write(fd, "hello\n", sizeof("hello\n")),
        sizeof("hello\n")
    );
    fsync(fd);
}

/* Register the test suite */
uk_testsuite_register(vfscore_stat_testsuite, NULL);
```

We will be using a simple app without any main function to run the testsuite, the output should be similar with:

```console
test: vfscore_stat_testsuite->vfscore_test_newfile
    :  expected `ret` to be 0 but was 0 ................................................ [ PASSED ]
    :  expected `ret` to be 0 but was 0 ................................................ [ PASSED ]
    :  expected `fd` to be greater than 2 but was 3 .................................... [ PASSED ]
    :  expected `write(fd, "hello\n", sizeof("hello\n"))` to be 7 but was 7 ............ [ PASSED ]
```

### 04. Add a Test Suite for `nolibc`

Add a new test suite for nolibc with four test cases in it.
You can use any POSIX function from nolibc for this task.

> Feel free to look over the [documentation](https://github.com/lancs-net/unikraft/blob/nderjung/uktest/lib/uktest/include/uk/test.h) to write more complex tests.

Reviewer: ```suggestion
Feel free to look over the [documentation](https://github.com/unikraft/unikraft/blob/staging/lib/uktest/README.md) to write more complex tests.
```

* **Branch coverage** is the percentage of branches executed during the testing (e.g. if or while).
* **Path coverage** is the percentage of paths executed during the testing.

We'll now go briefly over two other validation techniques, fuzzing and symbolic execution.

### Fuzzing

**Fuzzing** or fuzz testing is an automated software testing technique that involves providing invalid, unexpected, or random data as inputs to a computer program.
The program is then monitored for exceptions such as crashes, failing built-in code assertions, or potential memory leaks.

The most popular OS fuzzers are [`kAFL`](https://github.com/IntelLabs/kAFL) and [`syzkaller`](https://github.com/google/syzkaller), but research in this area is very active.

### Symbolic Execution

As per Wikipedia, **symbolic execution** is a means of analyzing a program to determine what inputs cause each part of a program to execute.
An interpreter follows the program, assuming symbolic values for inputs rather than obtaining actual inputs as normal execution of the program would.
An example of a program being symbolically executed can be seen in the figure below:

<Image

>       src="/imgs/symbex.png"

Reviewer: ```suggestion
      src="/images/symbex.png"
```

KUnit is not the only tool used for testing Linux, there are tens of tools used to test Linux at any time:

* Test suites
  * [Linux Test Project](https://github.com/linux-test-project/ltp) is a collection of tools
  * Static code analyzers ([`Coverity`](https://scan.coverity.com/), [`Coccinelle`](https://coccinelle.gitlabpages.inria.fr/website/), [`smatch`](https://lwn.net/Articles/691882/), [`sparse`](https://sparse.docs.kernel.org/en/latest/))
  * Module tests (KUnit)
  * Fuzzing tools ([`Trinity`](https://github.com/kernelslacker/trinity), [`Syzkaller`](https://github.com/google/syzkaller))
  * Subsystem tests
* Automatic testing
  * `kisskb`
  * [`0Day`](https://github.com/0day-ci/linux)
  * [`kernelci`](https://foundation.kernelci.org/)
  * `Kerneltests`

In the figure below, we can see that as more and better tools were developed we saw an increase in reported vulnerabilities.
There was a peak in 2017, after which a steady decrease which may be caused by the amount of tools used to verify patches before being upstreamed.

<Image

>       src="/imgs/linux_vulnerabilities.png"

Reviewer: ```suggestion
      src="/images/linux_vulnerabilities.png"
```
